### PR TITLE
Migration cancellation (T96429)

### PR DESCRIPTION
**Pull request by bgerstle**
#### Add cancellation functionality to the data migration
The migration now checks at various points if it was cancelled, and returns after setting its `finished` flag to `YES`.

#### Add cancel button to data migration view
When pressed, it prompts the user if they want to cancel the migration, and in doing so, remove all their old data.  If they decide to cancel, it will cancel the migration, wait until it is `finished`, and then remove all old and partially-migrated data.

### Bonus round
I also added some error handling for fetch requests in the `OldDataSchemaMigrator`.  We were previously letting them fail silently, but it's probably better to return an error in this case anyway.

Proof-of-concept for migration cancellation UX:

cancel flow gif

### CODE REVIEW

**Comment by coreyfloyd** [comment]
:+1: preview looks good

**Comment by berndsi** [comment]
Does this delete all local data (incl. onboarding, login, etc.) or just history and saved pages?

**Comment by bgerstle** [comment]
@berndsi I need to write some tests to make sure threading stuff & all data is deleted, but the intent is that **all** data should be wiped.

**Comment by bgerstle** [comment]
@coreyfloyd just double checking, you had a look at the "return if cancelled" and async polling logic in `OldDataSchemaMigrator` and `DataMigratorProgressViewController`?

**Comment by bgerstle** [comment]
@berndsi not even sure what data we're storing w.r.t. on boarding. do you mean whether or not we should someone the "login/skip" prompt?

**Comment by berndsi** [comment]
@bgerstle Yes. Whether you show the onboarding screen is one aspect. In
Android we have even more onboarding tracker flags. Whether the user has
seen the ToC onboarding. But those are the simple ones for a user
perspective.
I'd be more concerned about the user having to log in again, change
language preferences, show images preference, ....

**Comment by bgerstle** [comment]
Let me investigate.. :shipit:

**Comment by vibhabamba** [comment]
@bgerstle looks pretty good.

2 Questions:

Looks like this happens only if you went to Saved or Recent instead of when you download the app? Is that accurate?

Can we change the copy to be more user friendly please?

**Comment by bgerstle** [comment]
@vibhabamba dev copy can *always* be changed!  I just wrote what came to mind, do you have any suggestions?

> Looks like this happens only if you went to Saved or Recent instead of when you download the app? Is that accurate?

Not sure if I understand the question, but the pre-launch alert asking users if they want to migrate their data only happens if they have legacy data.  They will have legacy data if they visited any pages (recent) or saved them.

**Comment by berndsi** [comment]
@bgerstle I think @vibhabamba means: When exactly do you get the prompt. Right after you start the app or when you select saved pages/history?

**Comment by bgerstle** [comment]
@berndsi I just walked @vibhabamba through the UX and she's going to give me some new copy for the prompts.

everyone else, please don't merge until I get updated copy from vibha.

:-1:

File: wikipedia/Data/OldDataSchemaMigrator.m [hunk 1]
```diff
@@ -59,6 +66,17 @@ + (void)removeOldData {
...
 }
 
+- (void)sendDelegateCallbackForError:(NSError*)error {
+    dispatch_block_t sendError = ^{
+        [self.progressDelegate oldDataSchema:self didFinishWithError:error];
+    };
+    if ([NSThread isMainThread]) {
+        sendError();
+    } else {
+        dispatch_async(dispatch_get_main_queue(), sendError);
```
Comment (coreyfloyd) [inline wikipedia/Data/OldDataSchemaMigrator.m]: sendError()? does the block execute otherwise?

Comment (bgerstle) [inline wikipedia/Data/OldDataSchemaMigrator.m]: @coreyfloyd it does, either inline on L74 (as you mentioned) or passed to `dispatch_async` on L76

**Pull request closed without merging**